javascript/testing_javascript/project_battleship.md

```diff
@@ -9,27 +9,31 @@ We have not yet discussed testing the appearance of a webpage, which requires a
 1. Create `Gameboard` class/factory.
-   1. Note that we have not yet created any User Interface.  We should know our code is coming together by running the tests.  You shouldn't be relying on `console.log` or DOM methods to make sure your code is doing what you expect it to.
+   1. Note that we have not yet created any User Interface. We should know our code is coming together by running the tests. You shouldn't be relying on `console.log` or DOM methods to make sure your code is doing what you expect it to.
    1. Gameboards should be able to place ships at specific coordinates by calling the ship factory function.
    1. Gameboards should have a `receiveAttack` function that takes a pair of coordinates, determines whether or not the attack hit a ship and then sends the 'hit' function to the correct ship, or records the coordinates of the missed shot.
    1. Gameboards should keep track of missed attacks so they can display them properly.
    1. Gameboards should be able to report whether or not all of their ships have been sunk.
-1. Create `Player`.
-   1. Players can take turns playing the game by attacking the enemy Gameboard.
-   1. The game is played against the computer, so make the 'computer' capable of making random plays.  The AI does not have to be smart, but it should know whether or not a given move is legal (i.e. it shouldn't shoot the same coordinate twice).
-1. Create the main game loop and a module for DOM interaction.
-   1. At _this_ point it is appropriate to begin crafting your User Interface.
-   1. The game loop should set up a new game by creating Players and Gameboards.  For now just populate each Gameboard with predetermined coordinates.  You are going to implement a system for allowing players to place their ships later.
-   1. We'll leave the HTML implementation up to you for now, but you should display both the player's boards and render them using information from the Gameboard class/factory.
-      1. You need methods to render the gameboards and to take user input for attacking.  For attacks, let the user click on a coordinate in the enemy Gameboard.
-   1. The game loop should step through the game turn by turn using only methods from other objects.  If at any point you are tempted to write a new function inside the game loop, step back and figure out which class or module that function should belong to.
-   1. Create conditions so that the game ends once one player's ships have all been sunk.  This function is appropriate for the Game module.
+1. Create `Player`
+    1. There will be two types of players in the game, 'real' players and 'computer' players.
+    1. Each player object should contain it's own gameboard.
+1. Import your classes/factories into another file, and drive the game using event listeners to interact with your objects. Create a module that helps you manage actions that should happen in the DOM
```

Comment:
> I think the initial wording can be misinterpreted to mean to have one module for event listeners, and a completely separate module for DOM stuff (counter-intuitive).

I always interpreted it that way. I figured the intention was for learners to create a DOM module and put things like their `renderBoard` methods in it. 

Maybe we can just move that last sentence into it's own separate step?

```suggestion
1. Create a module to help you manage actions that that should happen in the DOM
    1. Leave this empty for now, you'll add methods to this module as you work through the game-driving code. 
1. Import your classes/factories into another module, and drive the game using event listeners to interact with your objects. 
```

Reply:
See, to me event listeners seem like they would be better off in that DOM module rather than separate, as they're also concerned with interacting with the DOM and simply an interface for calling Player methods, for example. And that perhaps there may even be room for a more defined "GameController" module.

I think this is one of the areas where it might therefore benefit to *not* be as prescriptive, as I've seen plenty of approaches in the community to handle this specific part.

Much of this project's purpose is to practise testing/TDD, so prescribing more about some of the early objects makes sense to get people started. But later aspects, I think can be left to the learner and we should probably try to keep to just clarifying the "game loop" confusion by avoiding that terminology and emphasising an event-driven nature. As that feels like more of the original issue at hand, and discussing the details of more prescription here feels a bit like a side extra?

Reply:
> See, to me event listeners seem like they would be better off in that DOM module rather than separate, as they're also concerned with interacting with the DOM and simply an interface for calling Player methods, for example. And that perhaps there may even be room for a more defined "GameController" module.

If you look at Bender's `ScreenController` pattern- it's pretty much just a "MVC" pattern, but with the "view" functions stuffed inside. If we're equating how learners should implement the flow of the game to something like that, then I think it makes a lot of sense to introduce a DOM/View module for learners to use, since it's possible the way learners interact with the DOM could be a bit more complicated here. 

With that said, I'll agree that it may not be required. Your suggestion would probably be good enough for learners to get the game running without too much hassle. 

But, I strongly believe with the code I've seen that keeping a sharp distinction between the game-driving code and a module that helps learners handle the View/DOM would help learners a lot. TOP doesn't really cover anything related to separating these types of concerns, so I think being less prescriptive with it has been bit of a bane for learners in this project. Perhaps it might be better though, with addition of the rest of what's been proposed here.

Encouraging learners to extract simple DOM functions to another module might:
- Encourage writing functions for simple DOM interactions in the first place
- Discourage functions that manipulate the DOM from becoming too coupled with other code in the game-driving module (they might have to actually write nice functions with like- arguments and stuff... not just use scope...)
- Encourage learners to streamline the game-driving module, so that it acts more like a MVC controller and just relays messages to various objects- including the DOM/View module. The current wording in this PR supports this notion nicely already.

I don't think this is a sticking issue for me. I can live moving on with your suggestion, it would just requiring removing a few more words, I think:
```md
    - You'll need methods to render each player's Gameboard, so put them in an appropriate module.
```
```md
    - You'll need methods to render each player's Gameboard.
```

Reply:
I think we may be misunderstanding each other a little :sweat_smile: 

I'm totally in agreement with separating DOM responsibilities from game flow responsibilities. What I meant was a little confusion about having event listeners in the game flow module and having DOM render/update methods (for example) in the DOM module.

Personally, I'd create a GameController that handles the game flow, and put *all* DOM stuff (including event listeners) in the DOM module, just like the "Building a House" Connect4 example. Since the event listeners are just an interface for the user to call certain methods.

I don't think the way I said that before was the clearest, apologies, and I think I misunderstood part of your intentions too XD

Since I think the original issue about "game loop" came about because of the word "loop", which often led to some whacky `while (true)` and Promise/async funkiness, what say you if we find some middle-ground between the current and proposed wording for that part?

e.g. `Create a module for the main game flow, and a module for DOM interaction.` (or "for the flow of the game" or something among those lines), while also having a bit about using event listeners?

At a glance, to me that feels like a good "not *too* prescriptive way of saying "make this event driven, separate concerns, perhaps remember TTT?". I'd personally find it pretty impressive to interpret this sort of framing as "infinite loop and async stuff", if that makes sense.

Reply:
Again, while the actual specifics of this part aren't really a sticking point for me either, I just want to make sure we're on the same page about the intent and scope of the instructions at least. We can always revisit the smaller details if community feedback/observations justify it.

Reply:
> Personally, I'd create a GameController that handles the game flow, and put all DOM stuff (including event listeners) in the DOM module, just like the "Building a House" Connect4 example. Since the event listeners are just an interface for the user to call certain methods.

Got it. I think that's valid too, I just don't think creating a `GameController` is what the current instructions lead learners to do (whether it was the intention or not). 

I'm not sure if you're saying:
- That's how you'd currently interpret the proposed instructions (and that's a problem), or 
- That's how you feel learners *should go about it* (and the current instructions contradict with that).

Within the framework of the current instructions on the website, it just makes a lot of sense to imagine the intention was to something like the `ScreenController` that directly interacted with independent player objects (no `GameController` middleman)- so that's kind of the direction I was aiming for. It's simple, and it works within the framework we already have.

> put all DOM stuff (including event listeners) in the DOM module, just like the "Building a House" Connect4 example.

In both situations (with and without a middleman `GameController` to run the game logic), I feel it makes sense to split that DOM module you mentioned up into independent controller/view modules at a certain size. That's what what my proposed changes were trying to suggest learners to do. Depending on how learners implement things, the amount of view functions could likely reach it.

Reply:
I'm unsure how to word this in a good way to suggest *both* possibilities (with a `GameController` middleman or without, if that's what you're trying to do. 

I think it's best to either just suggests users create a `GameController` to manage the game without any DOM stuff, or just not acknowledge the possibility at all to avoid confusion.

Reply:
> I'm not sure if you're saying:

More just saying how I'd personally do things based on the current instructions, which I don't think the current instructions contradict but are open enough to interpretation. But that relies on my reading of "main game loop" as "game flow", whereby I'd then choose to make a separate `GameController` (which others may not decide to do).

I think then let's go with what proposed originally (just add the `.` at the end of that line). I can see where you're coming from with that and think it does ultimately address the original issue brought up. If it would cause any further/new issues, we'll likely observe that in the community, since we're both active with helping on that front.

Reply:
> More just saying how I'd personally do things based on the current instructions, which I don't think the current instructions contradict but are open enough to interpretation.

Fair enough. I think finding a good way to clarify things for learners and make it more open to *either* interpretation might be rough. I'll think on it a little too. I can see how the current instructions are vague enough to support it.

I think it might be good to keep in mind that not everyone reads through the "Building a House from the Inside Out" article, or takes it super to heart carrying on in the curriculum. It's only recently been more prominent in the curriculum, and it's still really just a suggestion in TTT. 

It might be a bit much to *expect* learners to create something like a `GameController` to encapsulate game logic unless the instructions lean into it.

Reply:
That's a fair point about the article. Let's go with your original proposal, which I think still solves the main problem at hand. If it will cause problems, then we'll find out in due course and can address them as necessary but for now, I think we can go ahead :muscle: